crates/automations/src/server.rs

```diff
@@ -0,0 +1,235 @@
+    pub const fn new() -> Self {
+        Self(Vec::new())
+    }
+
+    /// Register an Executor to be served by this Server.
+    pub fn register<E: Executor>(mut self, executor: E) -> Self {
+        let index = match self
+            .0
+            .binary_search_by_key(&E::TASK_TYPE, |entry| entry.task_type())
+        {
+            Ok(_index) => panic!("an Executor for {:?} is already registered", E::TASK_TYPE),
+            Err(index) => index,
+        };
+
+        self.0.insert(index, Arc::new(executor));
+        self
+    }
+
+    /// Serve this Server until signaled to stop by `shutdown`.
+    pub async fn serve(
+        self,
+        permits: u32,
+        pool: sqlx::PgPool,
+        dequeue_interval: std::time::Duration,
+        heartbeat_timeout: std::time::Duration,
+        shutdown: impl std::future::Future<Output = ()>,
+    ) {
+        serve(
+            self,
+            permits,
+            pool,
+            dequeue_interval,
+            heartbeat_timeout,
+            shutdown,
+        )
+        .await
+    }
+}
+
+pub struct ReadyTask {
+    pub executor: Arc<dyn executors::ObjSafe>,
+    pub permit: tokio::sync::OwnedSemaphorePermit,
+    pub pool: sqlx::PgPool,
+    pub task: DequeuedTask,
+}
+
+pub struct DequeuedTask {
+    pub id: models::Id,
+    pub type_: TaskType,
+    pub parent_id: Option<models::Id>,
+    pub inbox: Option<Vec<SqlJson<(models::Id, Option<BoxedRaw>)>>>,
+    pub state: Option<SqlJson<BoxedRaw>>,
+    pub last_heartbeat: String,
+}
+
+pub async fn serve(
+    executors: Server,
+    permits: u32,
+    pool: sqlx::PgPool,
+    dequeue_interval: std::time::Duration,
+    heartbeat_timeout: std::time::Duration,
+    shutdown: impl std::future::Future<Output = ()>,
+) {
+    let semaphore = Arc::new(tokio::sync::Semaphore::new(permits as usize));
+
+    // Use Box::pin to ensure we can fullly drop `ready_tasks` later,
+    // as it may hold `semaphore` permits.
+    let mut ready_tasks = Box::pin(ready_tasks(
+        executors,
+        pool.clone(),
+        dequeue_interval,
+        heartbeat_timeout,
+        semaphore.clone(),
+    ));
+    tokio::pin!(shutdown);
+
+    // Poll for ready tasks and start them until `shutdown` is signaled.
+    while let Some(ready_tasks) = tokio::select! {
+        ready = ready_tasks.next() => ready,
+        () = &mut shutdown => None,
+    } {
+        let ready_tasks: Vec<ReadyTask> = match ready_tasks {
+            Ok(tasks) => tasks,
+            Err(err) => {
+                tracing::error!(?err, "failed to poll for tasks (will retry)");
+                Vec::new()
+            }
+        };
+
+        for ready in ready_tasks {
+            tokio::spawn(async move {
+                let (task_id, task_type, parent_id) =
+                    (ready.task.id, ready.task.type_, ready.task.parent_id);
+
+                if let Err(err) = executors::poll_task(ready, heartbeat_timeout).await {
+                    tracing::warn!(
```

Comment:
I wonder if we might want to make these errors more visible, perhaps by persisting them on the tasks table? I'm supposing that the idea is we should do error handling within the `poll` function, and this is really only for errors that we didn't otherwise catch or anticipate, which seems reasonable. Just want to check whether we're on the same page there, that the `state` ought to expose error message if the task is doing its own error handling.

Reply:
Exactly -- the place I got to was tasks are free to handled expected errors in any way they like, including sending them as messages or modeling them in task state, but they're not an "error" from this crate's perspective.